cmd/recorder/recorder.go

```diff
@@ -100,115 +102,119 @@ func (rec *Recorder) runBrowser(recURL string) error {
 	}
 
 	allocCtx, _ := chromedp.NewExecAllocator(context.Background(), opts...)
-	ctx, _ := chromedp.NewContext(allocCtx, contextOpts...)
-
-	chromedp.ListenTarget(ctx, func(ev interface{}) {
-		switch ev := ev.(type) {
-		case *cruntime.EventExceptionThrown:
-			slog.Error("chrome exception", slog.String("err", ev.ExceptionDetails.Text))
-			if ev.ExceptionDetails.Exception != nil {
-				slog.Error("chrome exception", slog.String("err", ev.ExceptionDetails.Exception.Description))
-			}
-		case *cruntime.EventConsoleAPICalled:
-			args := make([]string, 0, len(ev.Args))
-			for _, arg := range ev.Args {
-				var val interface{}
-				var str string
-				if len(arg.Value) > 0 {
-					err := json.Unmarshal(arg.Value, &val)
-					if err != nil {
-						slog.Error("failed to unmarshal", slog.String("err", err.Error()))
-						continue
-					}
-					str = fmt.Sprintf("%+v", val)
-				} else {
-					str = arg.Description
-				}
-				args = append(args, str)
-			}
-
-			str := fmt.Sprintf("chrome console %s %s", ev.Type.String(), strings.Join(args, " "))
 
-			slog.Debug(sanitizeConsoleLog(str))
-		}
-	})
-
-	if err := chromedp.Run(ctx, chromedp.Navigate(recURL)); err != nil {
-		return fmt.Errorf("failed to run chromedp: %w", err)
-	}
-
-	defer func() {
+	var ctx context.Context
+	cleanup := func() {
 		tctx, cancelCtx := context.WithTimeout(ctx, stopTimeout)
+		defer cancelCtx()
 		// graceful cancel
 		if err := chromedp.Cancel(tctx); err != nil {
 			slog.Error("failed to cancel context", slog.String("err", err.Error()))
 		}
-		cancelCtx()
-		close(rec.stoppedCh)
-	}()
+	}
 
-	ticker := time.NewTicker(connCheckInterval)
-	defer ticker.Stop()
+	defer func() {
+		cleanup()
+		rec.stoppedCh <- rerr
+	}()
 
-	var connected bool
-	connectCheckExpr := "window.callsClient && window.callsClient.connected && !window.callsClient.closed"
 	for {
 		select {
 		case <-rec.stopCh:
-			slog.Info("stop signal received, shutting down browser")
-			return nil
-		case <-ticker.C:
-			if err := chromedp.Run(ctx,
-				chromedp.Evaluate(connectCheckExpr, &connected),
-			); err != nil {
-				slog.Error("failed to run chromedp", slog.String("err", err.Error()))
-				continue
-			}
-			if !connected {
-				slog.Debug("not connected to call yet")
-				continue
-			}
-
-			slog.Debug("connected to call")
-			close(rec.readyCh)
+			return fmt.Errorf("stop signal received while initializing client")
```

Comment:
That's needed because if the browser continuously fails to start (line 169) we still want to exit properly when the global `readyTimeout` expires. Otherwise we'd be stuck there indefinitely.